geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaProducerFeatureStore.scala

```diff
@@ -64,86 +64,83 @@ class KafkaProducerFeatureStore(entry: ContentEntry,
-      val ret = id
-      id += 1
-      s"$ret"
-    }
+  protected val msgEncoder = new KafkaGeoMessageEncoder(sft)
 
-    var toModify: Iterator[SimpleFeature] =
-      if(query == null) Iterator[SimpleFeature]()
-      else if(query.getFilter == null) Iterator.continually {
-        reuse.getIdentifier.setID(getNextId)
-        reuse
-      }
-      else query.getFilter match {
-        case ids: Id        =>
-          ids.getIDs.map(id => new ScalaSimpleFeature(id.toString, sft)).iterator
+  private [kafka] def write(sf: SimpleFeature): Unit = send(GeoMessage.createOrUpdate(sf))
 
-        case Filter.INCLUDE =>
-          Iterator.continually(new ScalaSimpleFeature("", sft))
-      }
+  private [kafka] def send(msg: GeoMessage): Unit = {
+    logger.debug("sending message: {}", msg)
+    producer.send(msgEncoder.encodeMessage(topic, msg))
+  }
 
-    def setIter(iter: Iterator[SimpleFeature]): Unit = {
-      toModify = iter
-    }
+  override def getFeatureType: SimpleFeatureType = sft
+  override def close(): Unit = {}
+}
 
-    var curFeature: SimpleFeature = null
-    override def getFeatureType: SimpleFeatureType = sft
+class KafkaFeatureWriterAppend(sft: SimpleFeatureType, producer: Producer[Array[Byte], Array[Byte]], topic: String)
+    extends KafkaFeatureWriter(sft, producer, topic) {
 
-    override def next(): SimpleFeature = {
-      curFeature = toModify.next()
-      curFeature
-    }
+  protected val reuse = new ScalaSimpleFeature("", sft)
+
+  protected def nextId: String = KafkaProducerFeatureStoreFactory.FeatureIds.getAndIncrement().toString
 
-    override def remove(): Unit = {
-      val msg = GeoMessage.delete(curFeature.getID)
-      curFeature = null
+  // always return false as we're appending, per the geotools API
+  override def hasNext: Boolean = false
 
-      send(msg)
+  override def next(): SimpleFeature = {
+    reuse.getIdentifier.setID(nextId)
+    reuse.getUserData.clear()
+    var i = 0
+    while (i < reuse.values.length) {
+      reuse.values(i) = null
+      i += 1
     }
+    reuse
+  }
 
-    override def write(): Unit = {
-      val msg = GeoMessage.createOrUpdate(curFeature)
-      curFeature = null
+  override def write(): Unit = write(reuse)
 
-      send(msg)
-    }
+  override def remove(): Unit = throw new NotImplementedError("Remove called on FeatureWriterAppend")
+}
 
-    override def hasNext: Boolean = toModify.hasNext
-    override def close(): Unit = {}
+class KafkaFeatureWriterModify(sft: SimpleFeatureType, producer: Producer[Array[Byte], Array[Byte]], topic: String, query: Query)
+    extends KafkaFeatureWriterAppend(sft, producer, topic) {
 
-    private def send(msg: GeoMessage): Unit = {
-      logger.debug("sending message: {}", msg)
-      producer.send(msgEncoder.encodeMessage(topic, msg))
-    }
+  private val ids = query.getFilter match {
+    case ids: Id => ids.getIDs.iterator
+    case _ => throw new NotImplementedError("Only modify by ID is supported")
```

Comment:
limitation of the design - when we 'modify' a feature, we're actually just sending it off on a kafka queue. With a normal feature writer, you can do a scan to return 'live' features that can then be modified - but the kafka producer feature store doesn't actually have access to any features. And the kafka consumer feature store, which does have access to the features, can't modify them.